### PR TITLE
Model inference for connector classifier on queries

### PR DESCRIPTION
**Pull request by shukant-danswer**
## Description

The integrates the connector classifier model in the search preprocessing pipeline. Instead of using a LLM for auto detecting the sources, we use our own model instead if `FEATURE_FLAG_USE_CONNECTOR_CLASSIFIER` is set in the environment.

The model is warmed up by the API if the feature flag is set.

## How Has This Been Tested?
Hard coded `auto_detect_source_filter` and tested search queries in Danswer's UI.

## Checklist:
- [ ] All of the automated tests pass
- [ ] All PR comments are addressed and marked resolved
- [ ] If there are migrations, they have been rebased to latest main
- [ ] If there are new dependencies, they are added to the requirements
- [ ] If there are new environment variables, they are added to all of the deployment methods
- [ ] If there are new APIs that don't require auth, they are added to PUBLIC_ENDPOINT_SPECS
- [ ] Docker images build and basic functionalities work
- [x] Author has done a final read through of the PR right before merge

### CODE REVIEW

File: backend/danswer/main.py [hunk 1]
```diff
@@ -344,6 +346,12 @@ async def lifespan(app: FastAPI) -> AsyncGenerator:
                 model_server_port=MODEL_SERVER_PORT,
             )
 
+        if os.getenv("FEATURE_FLAG_USE_CONNECTOR_CLASSIFIER"):
```
Comment (yuhongsun96) [inline backend/danswer/main.py]: Would throw this one in app_configs.py or chat_configs.py for consistency

File: backend/danswer/secondary_llm_flows/source_filter.py [hunk 2]
```diff
@@ -42,11 +44,36 @@ def _sample_document_sources(
         return random.sample(valid_sources, num_sample)
 
 
+def _sample_documents_using_custom_connector_classifier(
+    query: str,
+    valid_sources: list[DocumentSource],
+) -> list[DocumentSource] | None:
+    available_connectors = list(filter(
+        lambda conn: conn.lower() in query.lower(),
```
Comment (yuhongsun96) [inline backend/danswer/secondary_llm_flows/source_filter.py]: would also remove all whitespaces and punctuation just to be safe, in case some user types Git Hub or something

File: backend/shared_configs/configs.py [hunk 3]
```diff
@@ -14,9 +14,12 @@
 )
 
 # Danswer custom Deep Learning Models
+CONNECTOR_CLASSIFIER_MODEL_REPO = "ShukantP/danswer-connector-classifier"
+CONNECTOR_CLASSIFIER_MODEL_TAG = "3fa7999c89ae72bc3892448dbd3a77c429e49153"
```
Comment (yuhongsun96) [inline backend/shared_configs/configs.py]: instead of using a SHA let's use SemVer standards or something approximately like that, can do v1.0.0

File: backend/shared_configs/model_server_models.py [hunk 4]
```diff
@@ -7,6 +7,16 @@
 Embedding = list[float]
 
 
+class ConnectorClassificationRequest(BaseModel):
+    available_connectors: list[str]
+    query: str
+
+
+class ConnectorClassificationResponse(BaseModel):
```
Comment (yuhongsun96) [inline backend/shared_configs/model_server_models.py:15]: I think this should abstract away the filter boolean and only return the connectors to filter by if any

File: backend/model_server/custom_models.py [hunk 5]
```diff
@@ -59,6 +103,76 @@ def get_local_intent_model(
     return _INTENT_MODEL
 
 
+def tokenize_connector_classification_query(
+    connectors: list[str],
+    query: str,
+    tokenizer: PreTrainedTokenizer,
+    connector_token_end_id: int,
+):
```
Comment (yuhongsun96) [inline backend/model_server/custom_models.py]: Be sure to static type returns as well

File: backend/model_server/custom_models.py [hunk 6]
```diff
@@ -59,6 +103,76 @@ def get_local_intent_model(
...
+    input_ids = torch.tensor([tokenizer.cls_token_id], dtype=torch.long)
+    attention_mask = torch.tensor([1], dtype=torch.long)
+
+    for connector in connectors:
```
Comment (yuhongsun96) [inline backend/model_server/custom_models.py:128]: where are you adding the prefix tokens?

File: backend/model_server/danswer_torch_model.py [hunk 7]
```diff
@@ -66,3 +66,54 @@ def from_pretrained(cls, load_directory: str) -> "HybridClassifier":
...
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        attention_mask: torch.Tensor,
+    ):
+        hidden_states = self.distilbert(input_ids=input_ids, attention_mask=attention_mask).last_hidden_state
+
+        cls_position_id = input_ids == self.tokenizer.cls_token_id
+        cls_hidden_states = hidden_states[cls_position_id]
+        global_logits = self.connector_global_classifier(cls_hidden_states).view(-1)
```
Comment (yuhongsun96) [inline backend/model_server/danswer_torch_model.py:105]: I would have done this on just the leading CLS token specifically

File: backend/model_server/danswer_torch_model.py [hunk 8]
```diff
@@ -66,3 +66,54 @@ def from_pretrained(cls, load_directory: str) -> "HybridClassifier":
...
+    def __init__(self, config: DistilBertConfig) -> None:
+        super().__init__()
+
+        self.config = config
+        self.distilbert = DistilBertModel(config)
+        self.connector_global_classifier = nn.Linear(self.distilbert.config.dim, 1)
+        self.connector_match_classifier = nn.Linear(self.distilbert.config.dim, 1)
+        self.tokenizer = DistilBertTokenizer.from_pretrained('distilbert-base-uncased')
+        self.dropout = nn.Dropout(self.distilbert.config.seq_classif_dropout)
+        self.connector_end_token_id = self.tokenizer.get_vocab()[self.config.connector_end_token]
```
Comment (yuhongsun96) [inline backend/model_server/danswer_torch_model.py]: would add a comment here on what this is

File: backend/model_server/custom_models.py [hunk 9]
```diff
@@ -59,6 +103,76 @@ def get_local_intent_model(
...
+    connector_classifier_tokenizer = get_connector_classifier_tokenizer()
+    connector_classifier = get_local_connector_classifier()
+
+    input_ids, attention_mask = tokenize_connector_classification_query(
+        ["GitHub"],
+        "danswer classifier query google doc",
+        connector_classifier_tokenizer,
+        connector_classifier.connector_end_token_id,
+    )
+    input_ids = input_ids.to(next(connector_classifier.parameters()).device)
```
Comment (yuhongsun96) [inline backend/model_server/custom_models.py]: slightly prefer for the model to know it's own device rather than checking the layers, see other recent PR

File: backend/model_server/danswer_torch_model.py [hunk 10]
```diff
@@ -66,3 +66,54 @@ def from_pretrained(cls, load_directory: str) -> "HybridClassifier":
...
+class ConnectorClassifier(nn.Module):
+    def __init__(self, config: DistilBertConfig) -> None:
+        super().__init__()
+
+        self.config = config
+        self.distilbert = DistilBertModel(config)
+        self.connector_global_classifier = nn.Linear(self.distilbert.config.dim, 1)
+        self.connector_match_classifier = nn.Linear(self.distilbert.config.dim, 1)
+        self.tokenizer = DistilBertTokenizer.from_pretrained('distilbert-base-uncased')
+        self.dropout = nn.Dropout(self.distilbert.config.seq_classif_dropout)
```
Comment (yuhongsun96) [inline backend/model_server/danswer_torch_model.py]: From my experience, when writing the inference version of models, people generally don't include this even as a layer definition

File: backend/model_server/danswer_torch_model.py [hunk 11]
```diff
@@ -66,3 +66,54 @@ def from_pretrained(cls, load_directory: str) -> "HybridClassifier":
...
+        connector_end_position_ids = (input_ids == self.connector_end_token_id)
+        connector_end_hidden_states = hidden_states[connector_end_position_ids]
+        connector_end_hidden_states = self.dropout(connector_end_hidden_states)
+        classifier_output = self.connector_match_classifier(connector_end_hidden_states)
+        classifier_confidence = torch.nn.functional.sigmoid(classifier_output).view(-1)
+
+        return global_confidence, classifier_confidence
+
+    @classmethod
+    def from_pretrained(cls, repo_dir: str) -> "ConnectorClassifier":
```
Comment (yuhongsun96) [inline backend/model_server/danswer_torch_model.py:116]: I would lock this to inference mode and turn off all gradient tracking

File: backend/model_server/main.py [hunk 12]
```diff
@@ -68,6 +69,7 @@ async def lifespan(app: FastAPI) -> AsyncGenerator:
     logger.info(f"Torch Threads: {torch.get_num_threads()}")
 
     if not INDEXING_ONLY:
+        # warm_up_connector_classifier_model()
```
Comment (yuhongsun96) [inline backend/model_server/main.py]: nit: rm this and the import and the function definition, let's not have bloat in the code. We will never warm up from the model server because even if this is always used, it's only applicable for Slack so we don't want to warm up this model on server startup regardless if we remove the feature flag.

File: backend/model_server/danswer_torch_model.py [hunk 13]
```diff
@@ -66,3 +66,54 @@ def from_pretrained(cls, load_directory: str) -> "HybridClassifier":
...
+            os.path.join(repo_dir, "pytorch_model.pt"),
+            map_location=device,
+            weights_only=True,
+        )
+
+        model = cls(config)
+        model.load_state_dict(state_dict)
+        model.to(device)
+
+        return model
```
Comment (yuhongsun96) [inline backend/model_server/danswer_torch_model.py]: nit: newline

File: backend/model_server/custom_models.py [hunk 14]
```diff
@@ -59,6 +103,76 @@ def get_local_intent_model(
     return _INTENT_MODEL
 
 
+def tokenize_connector_classification_query(
```
Comment (yuhongsun96) [inline backend/model_server/custom_models.py:113]: generally with these types of functions, it's good to give a comment of what it might look like, write out the natural language representations of the tokens as a list or something along with the mask etc.

File: backend/model_server/danswer_torch_model.py (hunk 13 above)
Comment (ghost) [reply backend/model_server/danswer_torch_model.py]: I typically like separating blocks of code with newlines; removed it here, but it's pretty similar in the `HybridClassifier`

File: backend/model_server/danswer_torch_model.py (hunk 10 above)
Comment (ghost) [reply backend/model_server/danswer_torch_model.py]: Good pt. The `HybridClassifier` can also be cleaned up in this regard.

File: backend/danswer/secondary_llm_flows/source_filter.py (hunk 2 above)
Comment (ghost) [reply backend/danswer/secondary_llm_flows/source_filter.py]: It depends if you want to lean towards filtering vs not filtering. It could backfire if the user asks "sending mail" and that becomes "gmail"

File: backend/model_server/custom_models.py (hunk 6 above)
Comment (ghost) [reply backend/model_server/custom_models.py:128]: after the names of each connector, L131 now

File: backend/model_server/custom_models.py (hunk 9 above)
Comment (ghost) [reply backend/model_server/custom_models.py]: I can see the appeal of that - updated.

File: backend/danswer/secondary_llm_flows/source_filter.py (hunk 2 above)
Comment (yuhongsun96) [reply backend/danswer/secondary_llm_flows/source_filter.py]: ya... but the model should handle it! I see this step as more of a complexity reduction but we can lean towards allowing connectors to be passed to the model if it's a case like this

File: backend/model_server/danswer_torch_model.py (hunk 10 above)
Comment (yuhongsun96) [reply backend/model_server/danswer_torch_model.py]: ah ya, feel free to toss that in!

File: backend/danswer/secondary_llm_flows/source_filter.py (hunk 2 above)
Comment (ghost) [reply backend/danswer/secondary_llm_flows/source_filter.py]: true - updated!

File: backend/model_server/danswer_torch_model.py (hunk 10 above)
Comment (ghost) [reply backend/model_server/danswer_torch_model.py]: done!
